protocol.md

```diff
@@ -0,0 +1,338 @@
+```
+
+As you can see, the message is an array and each index value has its meaning:
+* 0: `c4` is a decoder handler identification for the current payload format. Identifications allow to register handler for many different types of messages
+* 1: array which items correspond to the described payload fields above
+
+For more details regarding serialization and deserialization please consult [transit format](https://github.com/cognitect/transit-format) specification.
+
+## Content types
+
+Here is a list of supported content types:
+* `text/plain` identifies a message which content is a plain text
+
+TODO: are there other types? How is it useful? Sending Eth requests?
+
+## Message types
+
+Here is a list of supported message types:
+* `public-group-user-message` is a message to the public group
+* `user-message` is a private message.
+
+TODO: how is it useful? What are other supported message types?
+
+## Clock vs Timestamp And Message Ordering
+
+`timestamp` is Unix time calculated when the message is created (TODO: does it come from Whisper or the system?)
+
+`clock` is calculated using the algorithm of [Lamport timestamps](https://en.wikipedia.org/wiki/Lamport_timestamps). When there are messages available in a chat, `clock`'s value is calculated as `last-message-clock-value + 1`. If there are no messages, `clock` is initialized with `timestamp`'s value.
+
+`clock` value is used for the message ordering. Due to the used algorithm, distributed nature of the system and physics, we achieve casual ordering which might produce counterintuitive results in some edge cases. For example, when one joins a public chat and sends a message before receiving the exist messages, their message `clock` value might be lower and the message will end up in the past when the historical messages are fetched.
+
+## Quoting
+
+TBD
+
+# Whisper adapter
+
+Whisper has been chosen as an messages exchange protocol because it was designed as an off-chain communication layer for the Ethereum nodes. It supports e2e encryption and uses epidemic spread to route data to all members of the network. It also provides [darkness to some extent](https://github.com/ethereum/go-ethereum/wiki/Achieving-Darkness).
+
+However, Whisper was not designed to handle huge number of messages and real-time communication. These are the tradeoffs that we accepted when implementing this protocol.
+
+This protocol can either work using a Whisper service which requires a protocol implementation to run in the same process as a Whisper node as well as using a Whisper client which might run as a separate Ethereum node and communicate through IPC or WebSocket.
+
+There is some tight coupling between the payload and Whisper:
+* Whisper message topic depends on the actual message type (see [Topic](#topic))
+* Whisper message uses a different key (asymmetric or symmetric) depending on the actual message type (see [Keys management](#keys-management))
```

Comment:
This should refer back to more precise crypto things used

Reply:
I would link to http://eips.ethereum.org/EIPS/eip-627. I can't find anything more detailed in Whisper docs. We don't control cryptographic algorithms here.

Reply:
That's because the Whisper spec isn't good enough. It should be possible to implement e.g. a client in Nim from this, and if there's a problem with some primitive we use (say hash fn, or slow KDF or whatever) it should be very clear what we are relying on. E.g. see https://code.briarproject.org/briar/briar-spec/blob/master/protocols/BTP.md and Signal/Whatsapp papers

Reply:
In my opinion, before starting implementing this protocol, a client should have an implementation of Whisper V6 spec ready. We can discuss this later as it does not look like the most important part.